L1Trigger/L1TMuonEndCapPhase2/src/DAQ/GEMTPSelector.cc

```diff
@@ -0,0 +1,110 @@
+
+#include "L1Trigger/L1TMuonEndCapPhase2/interface/DAQ/GEMTPSelector.h"
+
+using namespace emtf::phase2;
+
+GEMTPSelector::GEMTPSelector(const EMTFContext& context, const int& endcap, const int& sector)
+    : context_(context), endcap_(endcap), sector_(sector) {
+  // Do Nothing
+}
+
+GEMTPSelector::~GEMTPSelector() {
+  // Do Nothing
+}
+
+void GEMTPSelector::select(const TriggerPrimitive& tp, TPInfo tp_info, ILinkTPCMap& ilink_tpc_map) const {
+  emtf_assert(tp.subsystem() == L1TMuon::kGEM);
+
+  // Map GEM trigger primitives to input links
+  int ilink = get_input_link(tp, tp_info);  // Returns GEM "link" index
+
+  // Short-Circuit: Link not found (ilink = -1)
+  if (ilink < 0) {
+    return;
+  }
+
+  ilink_tpc_map[ilink].emplace_back(tp, tp_info);
+}
+
+// ===========================================================================
+// Utils
+// ===========================================================================
+int GEMTPSelector::get_input_link(const TriggerPrimitive& tp, TPInfo& tp_info) const {
+  int ilink = -1;
+
+  // Unpack detector info
+  const int tp_endcap = tp_info.endcap;
+  const int tp_sector = tp_info.sector;
+  const int tp_subsector = tp_info.subsector;
+  const int tp_station = tp_info.station;
+  const int tp_ring = tp_info.ring;
+  const int tp_csc_id = tp_info.csc_id;
+
+  // Find selection type
+  auto tp_selection = TPSelection::kNone;
+
+  if (csc::is_in_sector(endcap_, sector_, tp_endcap, tp_sector)) {
+    tp_selection = TPSelection::kNative;
+  } else if (CONFIG.include_neighbor_en_ &&
+             csc::is_in_neighbor_sector(endcap_, sector_, tp_endcap, tp_sector, tp_subsector, tp_station, tp_csc_id)) {
+    tp_selection = TPSelection::kNeighbor;
+  } else {  // Short-Circuit: tp_selection = TPSelection::kNone
+    return ilink;
+  }
+
+  // Get chamber input link for this sector processor
+  ilink = calculate_input_link(tp_subsector, tp_station, tp_ring, tp_csc_id, tp_selection);
+
+  // Add selection info
+  tp_info.ilink = ilink;
+  tp_info.selection = tp_selection;
+
+  return ilink;
+}
+
+int GEMTPSelector::calculate_input_link(const int& tp_subsector,
+                                        const int& tp_station,
+                                        const int& tp_ring,
+                                        const int& tp_csc_id,
+                                        const TPSelection& tp_selection) const {
+  int ilink = -1;
+
+  // Links
+  // RE1,2,3,4 + GE1,2        : 54..71, 72..80, 81..89, 90..98
+  // RE1,2,3,4 + GE1,2 (N)    : 99..101, 102..103, 104..105, 106..107
+
+  if (tp_selection == TPSelection::kNative) {
+    const int ilink_offset = 54;
+
+    if (tp_station == 1) {
+      ilink = ilink_offset + (tp_subsector - 1) * 9 + (tp_csc_id - 1);
+    } else {
+      ilink = ilink_offset + tp_station * 9 + (tp_csc_id - 1);
+    }
+
+    emtf_assert((54 <= ilink) && (ilink < 99));
+  } else {
+    const int ilink_offset = 99;
```

Comment:
Should the literal 54's and 99 be replaced by (explained) constants?

Reply:
These constants are shown in the input link order above.